xonsh/commands_cache.py

```diff
@@ -184,9 +183,62 @@ def predict_threadable(self, cmd):
                 pre, ext = os.path.splitext(name)
                 if pre in predictors:
                     predictors[name] = predictors[pre]
+        if name not in predictors:
+            predictors[name] = self.default_predictor(name, cmd[0])
         predictor = predictors[name]
         return predictor(cmd[1:])
 
+    #
+    # Background Predictors (as methods)
+    #
+
+    def default_predictor(self, name, cmd0):
+        if ON_POSIX:
+            return self.default_predictor_readbin(name, cmd0,
+                                                  timeout=.1,
+                                                  failure=predict_true)
+        else:
+            return predict_true
+
+    def default_predictor_readbin(self, name, cmd0, timeout, failure):
+        """Make a defautt predictor by
+        analyzing the content of the binary. Should only works on POSIX.
+        Return failure if the analysis fails.
+        """
+
+        fname = cmd0 if os.path.isabs(cmd0) else self.locate_binary(name)
+        if fname is None:
+            return failure
+
+        try:
+            fd = os.open(fname, os.O_RDONLY | os.O_NONBLOCK)
+        except:
+            return failure  # opening error
+
+        search_for = {
+            (b'ncurses',): [False, ],
+            (b'isatty', b'tcgetattr', b'tcsetattr'): [False, False, False],
+        }
+        tstart = os.times()[4]
+        block = b''
+        while os.times()[4] < tstart + timeout:
+            previous_block = block
+            block = os.read(fd, 2048)
+            if len(block) == 0:
+                os.close(fd)
+                return predict_true  # no keys of search_for found
+            analyzed_block = previous_block + block
+            for k, v in search_for.items():
+                for i in range(len(k)):
```

Comment:
Might be nice to have an `if v[i]: continue` block here so we don't recheck for keys we have already found.